### PR TITLE
Improve `sinkhorn_gibbs`

### PR DESCRIPTION
**Pull request by devmotion**
This PR improves `sinkhorn_gibbs` and fixes https://github.com/JuliaOptimalTransport/OptimalTransport.jl/issues/86.

### CODE REVIEW

**Comment by devmotion** [comment]
It would be great if you could comment on the PR, I think it would be good to fix the AD issue and to improve the sinkhorn implementation.

File: test/entropic.jl [hunk 1]
```diff
@@ -80,52 +103,51 @@ Random.seed!(100)
...
-            @test c_all ≈ c_pot rtol = 1e-6
-
-            γ_all = sinkhorn(μ[:, 1], ν, C, eps; maxiter=5_000)
-            γ_pot = [POT.sinkhorn(μ[:, 1], ν[:, i], C, eps; numItermax=5_000) for i in 1:d]
-            @test all([isapprox(γ_all[:, :, i], γ_pot[i]; rtol=1e-6) for i in 1:d])
-
-            γ_all = sinkhorn(μ, ν[:, 1], C, eps; maxiter=5_000)
-            γ_pot = [POT.sinkhorn(μ[:, i], ν[:, 1], C, eps; numItermax=5_000) for i in 1:d]
-            @test all([isapprox(γ_all[:, :, i], γ_pot[i]; rtol=1e-6) for i in 1:d])
+            # compute gradients with respect to source and target marginals separately and
```
Comment (zsteve) [inline test/entropic.jl:137]: I'll see if I can make a separate PR to check the correctness of these gradients.

File: src/OptimalTransport.jl [hunk 2]
```diff
@@ -87,43 +83,66 @@ function sinkhorn_gibbs(
...
+
+        # compute next iterate
+        u .= μ ./ Kv
+        mul!(v, K', u)
+        v .= ν ./ v
+        mul!(Kv, K, v)
+
+        # check source marginal
+        # always check convergence after the final iteration
+        if to_check_step <= 0 || iter == maxiter
```
Comment (zsteve) [inline src/OptimalTransport.jl:128]: is there a reason that we stop using `iter % check_step == 0`?

Comment (devmotion) [reply src/OptimalTransport.jl:128]: Yes, modulo operations are very expensive, the new version is a lot faster. A simple benchmark:
```julia
julia> using BenchmarkTools

julia> function f(n, m)
           x = 0
           for i in 1:n
              if i % m == 0
                  x += 1
              end
           end
           return x
       end
f (generic function with 1 method)

julia> function g(n, m)
           to_m = m
           x = 0
           for i in 1:n
               to_m -= 1
               if to_m == 0
                   x += 1
                   to_m = m
               end
           end
           return x
       end
g (generic function with 1 method)

julia> @benchmark f($(100_000), $(10))
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     267.902 μs (0.00% GC)
  median time:      273.221 μs (0.00% GC)
  mean time:        276.168 μs (0.00% GC)
  maximum time:     468.334 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1

julia> @benchmark g($(100_000), $(10))
BenchmarkTools.Trial: 
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     57.314 μs (0.00% GC)
  median time:      63.209 μs (0.00% GC)
  mean time:        63.485 μs (0.00% GC)
  maximum time:     95.818 μs (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     1
```

**Comment by devmotion** [comment]
What do you think, @zsteve, is it good or should anything be changed?

**Comment by davibarreira** [comment]
Sorry, I saw that my review was requested, but I haven't been able to do it yet. I still have to take a better look at the code. Feel free to merge without my input, cause I'll probably take a bit longer to review.

**Comment by zsteve** [comment]
> What do you think, @zsteve, is it good or should anything be changed?

This looks good!

**Pull request merged**